cpp/src/fil/infer.cu

```diff
@@ -531,45 +555,67 @@ template <int NITEMS, leaf_algo_t leaf_algo, bool cols_in_shmem,
+    int block_num_rows =
+      max(0, (int)min((long long)rows_per_block, params.num_rows - block_row0));
     const float* block_input = params.data + block_row0 * num_cols;
     if (cols_in_shmem) {
       // cache the row for all threads to reuse
-      size_t feature = 0;
+      // 2021: latest SMs still do not have >262KB of shared memory/block required to
+      // exceed the unsigned short
 #pragma unroll
-      for (feature = threadIdx.x; feature < num_input_rows * num_cols;
-           feature += blockDim.x)
-        sdata[feature] = block_input[feature];
+      for (unsigned short input_idx = threadIdx.x;
+           input_idx < block_num_rows * num_cols; input_idx += blockDim.x) {
+        // assuming here that sdata_stride == num_cols + 1
+        // then, idx / num_cols * sdata_stride + idx % num_cols == idx + idx / num_cols
+        unsigned short sdata_idx =
+          sdata_stride == num_cols
+            ? input_idx
+            : input_idx + input_idx / (unsigned short)num_cols;
+        sdata[sdata_idx] = block_input[input_idx];
+      }
 #pragma unroll
-      for (; feature < NITEMS * num_cols; feature += blockDim.x)
-        sdata[feature] = 0.0f;
+      for (int idx = block_num_rows * sdata_stride;
+           idx < rows_per_block * sdata_stride; idx += blockDim.x)
+        sdata[idx] = 0.0f;
     }
 
     tree_aggregator_t<NITEMS, leaf_algo> acc(
       params, (char*)sdata + params.cols_shmem_size(), sdata);
 
     __syncthreads();  // for both row cache init and acc init
-
-    // one block works on NITEMS rows and the whole forest
-    for (int j = threadIdx.x; j - threadIdx.x < forest.num_trees();
-         j += blockDim.x) {
-      /* j - threadIdx.x < forest.num_trees() is a necessary but block-uniform
-         condition for "j < forest.num_trees()". It lets use __syncthreads()
+    // one block works on NITEMS * threads_per_tree rows and the whole forest
+    // one thread works on NITEMS rows
+
+    int thread_tree0 = threadIdx.x >> params.log2_threads_per_tree;
+    int tree_stride = blockDim.x >> params.log2_threads_per_tree;
+    int thread_num_rows = max(0, min(NITEMS, block_num_rows - thread_row0));
+    for (int tree = thread_tree0; tree - thread_tree0 < forest.num_trees();
+         tree += tree_stride) {
+      /* tree - thread_tree0 < forest.num_trees() is a necessary but block-uniform
+         condition for "tree < forest.num_trees()". It lets use __syncthreads()
          and is made exact below.
+         Same with thread_num_rows > 0
       */
-      if (j < forest.num_trees()) {
-        acc.accumulate(infer_one_tree<NITEMS, leaf_output_t<leaf_algo>::T>(
-                         forest[j], cols_in_shmem ? sdata : block_input,
-                         num_cols, num_input_rows),
-                       j, num_input_rows);
+      if (tree < forest.num_trees() && thread_num_rows != 0) {
+        auto prediction = infer_one_tree<NITEMS, leaf_output_t<leaf_algo>::T>(
```

Comment:
Consider specifying the type of `prediction` explicitly. Feel free to use extra typedefs if you find them helpful.